src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -5790,73 +5985,115 @@ private bool CheckInterpolatedStringHandlerConversionEscape(BoundExpression expr
             var previousVisited = _visited;
             _visited = null;
 #endif
-            CheckValEscape(expression.Syntax, data.Construction, escapeFrom, escapeTo, checkingReceiver: false, diagnostics);
+            bool result = CheckValEscape(expression.Syntax, data.Construction, escapeFrom, escapeTo, checkingReceiver: false, diagnostics);
 #if DEBUG
             _visited = previousVisited;
 #endif
 
-            var arguments = ArrayBuilder<BoundExpression>.GetInstance();
-            GetInterpolatedStringHandlerArgumentsForEscape(expression, arguments);
+            // Narrow the scope for implicit calls which allow the receiver to capture refs from the arguments.
+            result = result && CheckValEscapeOfInterpolatedStringHandlerCalls(expression, escapeFrom, escapeTo, diagnostics);
 
-            bool result = true;
-            foreach (var argument in arguments)
+            return result;
+        }
+
+        private SafeContext GetValEscapeOfInterpolatedStringHandlerCalls(BoundExpression expression, SafeContext localScopeDepth)
+        {
+            SafeContext scope = SafeContext.CallingMethod;
+
+            while (true)
             {
-                if (!CheckValEscape(argument.Syntax, argument, escapeFrom, escapeTo, checkingReceiver: false, diagnostics))
+                switch (expression)
                 {
-                    result = false;
-                    break;
+                    case BoundBinaryOperator binary:
+                        scope = scope.Intersect(GetValEscapeOfInterpolatedStringHandlerCalls(binary.Right, localScopeDepth));
+                        expression = binary.Left;
+                        break;
+
+                    case BoundInterpolatedString interpolatedString:
+                        return scope.Intersect(getParts(interpolatedString, localScopeDepth));
+
+                    default:
+                        throw ExceptionUtilities.UnexpectedValue(expression.Kind);
                 }
             }
 
-            arguments.Free();
-            return result;
+            SafeContext getParts(BoundInterpolatedString interpolatedString, SafeContext localScopeDepth)
```

Comment:
If I see something named `getParts`, I expect it literally returns all the parts. Probably want to rename for clarity.
```suggestion
            SafeContext getPartsScope(BoundInterpolatedString interpolatedString, SafeContext localScopeDepth)
```